src/goto-programs/remove_calls_nobody.cpp

```diff
@@ -0,0 +1,116 @@
+/*******************************************************************\
+
+Module: Remove calls to functions without a body
+
+Author: Daniel Poetzl
+
+\*******************************************************************/
+
+/// \file
+/// Remove calls to functions without a body
+
+#include "remove_calls_nobody.h"
+
+void remove_calls_nobodyt::remove_call_nobody(
+  goto_programt &dest,
+  goto_programt::targett target,
+  const exprt &lhs,
+  const exprt::operandst &arguments)
+{
+  assert(target->is_function_call());
+  assert(!dest.empty());
+
+  goto_programt tmp;
+
+  // evaluate function arguments -- they might have
+  // pointer dereferencing or the like
+  forall_expr(it, arguments)
+  {
+    goto_programt::targett t=tmp.add_instruction();
+    t->make_other(code_expressiont(*it));
+    t->source_location=target->source_location;
+    t->function=target->function;
+  }
+
+  // return value
+  if(lhs.is_not_nil())
+  {
+    side_effect_expr_nondett rhs(lhs.type());
+    rhs.add_source_location()=target->source_location;
+
+    code_assignt code(lhs, rhs);
+    code.add_source_location()=target->source_location;
+
+    goto_programt::targett t=tmp.add_instruction(ASSIGN);
+    t->source_location=target->source_location;
+    t->function=target->function;
+    t->code.swap(code);
+  }
+
+  // kill call
+  target->type=LOCATION;
+  target->code.clear();
+  target++;
+
+  dest.destructive_insert(target, tmp);
+}
+
+void remove_calls_nobodyt::remove_calls_nobody(
```

Comment:
Instead of all being called `remove_calls_nobody` maybe this could be called `remove_no_body_calls_from_function`